src/rdpc/analysesProcessor.ts

```diff
@@ -0,0 +1,281 @@
+    state?: StreamState;
+  }
+): AsyncGenerator<Analysis[]> {
+  const chunkSize = config?.chunkSize || 1000;
+  const streamState: StreamState = {
+    currentPage: config?.state?.currentPage || 0,
+  };
+  while (true) {
+    const page = await fetchSeqExpAnalyses(
+      studyId,
+      url,
+      streamState.currentPage,
+      chunkSize
+    );
+    streamState.currentPage = streamState.currentPage + chunkSize;
+
+    if (page && page.length > 0) {
+      yield page;
+    } else {
+      break;
+    }
+  }
+};
+
+/**
+ * Extracts donor-run relation from analysis, this is done by the following steps:
+ * Creates inputAnalyses-run map for each donor by grouping runs by inputAnalyses,
+ * creates donor-inputAnalyses-run map by determining the latest run for each inputAnalyses,
+ * aggregates donor-inputAnalyses-run map by donor id.
+ * @param analyses RDPC analysis array
+ */
+export const toDonorCentric = (analyses: Analysis[]): DonorDocMap => {
+  const result = analyses.reduce<DonorDocMap>((acc, analysis) => {
+    const donorWithLatestRun = analysis.donors.reduce<DonorDocMap>(
+      (_acc, donor) => {
+        const inputAnalysesMap = _(analysis.runs)
+          .groupBy((run) => {
+            return (
+              HashCode(run.inputAnalyses[0].analysisId) +
+              HashCode(run.inputAnalyses[1]?.analysisId)
+            );
+          })
+          .value() as InputAnalysesRunMap;
+
+        Object.entries(inputAnalysesMap).forEach(([inputId, runs]) => {
+          const latestRun = getLatestRun(runs);
+          const run = latestRun === undefined ? [] : [latestRun];
+          const existingMap = _acc[donor.donorId];
+          _acc[donor.donorId] = {
+            ...existingMap,
+            [inputId]: run,
+          };
+        });
+
+        return _acc;
+      },
+      {}
+    );
+
+    // merge donor-inputAnalyses-run map by donorId, in case same donors appear under multiple analyses
+    Object.entries(donorWithLatestRun).forEach(
+      ([donorId, inputAnalysesMap]) => {
+        const existingMap = acc[donorId] ? acc[donorId] : {};
+        acc[donorId] = {
+          ...inputAnalysesMap,
+          ...existingMap,
+        };
+      }
+    );
+
+    return acc;
+  }, {});
+
+  return result;
+};
+
+const getLatestRun = (runs: Run[]): Run | undefined => {
+  return _(runs)
+    .sortBy(
+      (run) =>
+        (({ COMPLETE: 1, RUNNING: 2, EXECUTOR_ERROR: 3 } as {
+          [k: string]: number;
+        })[run.state])
+    )
+    .head();
+};
+
+export const mergeDonorMaps = (
+  mergedMap: DonorDocMap,
+  toMerge: DonorDocMap
+): DonorDocMap => {
+  Object.entries(toMerge).forEach(([donorId, inputAnalysesMap]) => {
+    const existingMap = mergedMap[donorId] ? mergedMap[donorId] : {};
+    const mergedInputAnalysesMap = {
+      ...existingMap,
+      ...inputAnalysesMap,
+    };
+    mergedMap[donorId] = { ...mergedInputAnalysesMap };
+  });
+
+  return mergedMap;
+};
+
+export const getAllMergedDonor = async (
```

Comment:
I think it reads better as `getAllRunsByAnalysesByDonors`... since what it's returning is basically
```
{
  [donorId: string]: {
    [analysisId: string]: Run[]
  }
}
```
which is a collection of `Runs` under a specific grouping